design-docs/modes.md

```diff
@@ -0,0 +1,638 @@
+# Summary
+
+This document describes a standardized approach for operators to select different code to run for different robot modes of operation and for programmers to easily write code that creates these selection options.
+
+A note on terminology: the word "mode" is used in this document to describe this functionality.  While this word overlaps with the more general "modes" of robot operation (e.g. disabled, periodic, and auto), it's natural to describe the robot as having "several autonomous modes," and "in autonomous mode" (without referring to a specific one) refers to the general robot mode.
+
```

Comment:
I think this confusion is probably a good enough reason for us to come up with another term than modes.

Reply:
"Routine" was about the only other option I came up with, unless we just want to stick with "OpMode" (which works, but might be confusing during the transition period, which is why I didn't use it).